### PR TITLE
[M] Updated refresher to avoid fetching unnecessary versioned entities (ENT-3845)

### PR DESCRIPTION
**Pull request by Ceiu**
- Updated the refresher infrastructure to add a 3rd processing stage
  which performs versioned entity resolution and persistence
  operations to avoid needing to prefetch unnecessary versioned
  entities and massively reducing the memory footprint of refresh
  and manifest import
- Updated the way nodes store links to other nodes to avoid needing
  to pass through entity mappers to visitors
- Rewrote several test suites to address the above changes and to
  improve test stability

### CODE REVIEW

**Review by Januson — COMMENTED** [review]
Changes look good. I have just a few suggestions.

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java [hunk 1]
```diff
@@ -171,13 +171,76 @@ private boolean clearedForDeletion(EntityNode<Content, ContentInfo> node) {
 
         // Otherwise, if the node is referenced by one or more parent nodes that are not being
         // deleted themselves, we should keep it.
-        for (EntityNode parent : node.getParentNodes()) {
-            if (parent.getNodeState() != NodeState.DELETED) {
-                return false;
+        return !node.getParentNodes()
+            .filter(elem -> elem.getNodeState() != NodeState.DELETED)
+            .findAny()
+            .isPresent​();
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java:177]: Fixed.

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java [hunk 2]
```diff
@@ -97,24 +103,19 @@ public ProductNodeVisitor(ProductCurator productCurator, OwnerProductCurator own
...
-            if (child.changed()) {
-                childrenUpdated = true;
-                break;
-            }
-        }
+        // Check if we need to make reference updates on this entity due to children updates
+        boolean childrenUpdated = node.getChildrenNodes()
+            .filter(EntityNode::changed)
+            .findAny()
+            .isPresent();
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java:118]: Fixed.

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java [hunk 3]
```diff
@@ -189,13 +182,82 @@ private boolean clearedForDeletion(EntityNode<Product, ProductInfo> node) {
 
         // Otherwise, if the node is referenced by one or more parent nodes that are not being
         // deleted themselves, we should keep it.
-        for (EntityNode parent : node.getParentNodes()) {
-            if (parent.getNodeState() != NodeState.DELETED) {
-                return false;
+        return !node.getParentNodes()
+            .filter(elem -> elem.getNodeState() != NodeState.DELETED)
+            .findAny()
+            .isPresent();
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java:188]: Fixed.

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java:177]: Yeah... I noticed that in a few other places, too. My IDE had a major update recently, so I wonder if that's causing these to show up.

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java [hunk 4]
```diff
@@ -171,13 +171,76 @@ private boolean clearedForDeletion(EntityNode<Content, ContentInfo> node) {
...
+            this.ownerContentEntities.add(new OwnerContent(node.getOwner(), node.getMergedEntity()));
+        }
+    }
+
+    /**
+     * Performs version resolution for the specified entity. If a local version of the entity
+     * already exists, this method returns the existing entity; otherwise, the provided entity
+     * is returned instead.
+     *
+     * @param owner
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java:215]: Fixed.

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java [hunk 5]
```diff
@@ -189,13 +182,82 @@ private boolean clearedForDeletion(EntityNode<Product, ProductInfo> node) {
...
+            this.ownerProductEntities.add(new OwnerProduct(node.getOwner(), node.getMergedEntity()));
+        }
+    }
+
+    /**
+     * Performs version resolution for the specified entity. If a local version of the entity
+     * already exists, this method returns the existing entity; otherwise, the provided entity
+     * is returned instead.
+     *
+     * @param owner
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java:226]: Fixed.

**Review by Januson — APPROVED** [review]

**Pull request merged**